### ISSUE TITLE
[SPIKE] Develop solution for improving the send step

### ISSUE THREAD

**Issue by chris-kuryak** [CDCgov/prime-reportstream#16472]
## Problem statement
The Platform team has identified that the Send step may be running inefficiently and slowing down how quickly we can send data to our receivers.

## What you need to know

- See existing guidance here from Platform: #15122
- Look at `SendFunction.kt` 
- Add metrics to these transport types:
- SFTP
- BLOBSTORE
- SOAP
- REST

## Acceptance criteria

- [ ] #15122 is updated to reflect the strategy to improve the Send step with Acceptance Criteria and reviewed with Product and Tech Lead
OR
- [ ] #15122 is deleted from the board (closed) and new ticket(s) created (and linked in the comments) to make improvements to the Send step with Acceptance Criteria and reviewed with Product and Tech Lead

**Comment by chris-kuryak** [comment]
@arnejduranovic Is this something Platform could take on so we can focus on onboardings?

**Comment by lucero-v** [comment]
Covered Victor on higher-priority tasks while he was out and assisted teammates with tasks.

**Comment by oslynn** [comment]
This looks like a lot of work for one ticket.  Perhaps, we can break it into couple tickets.  I can help work on this as well.

**Comment by chris-kuryak** [comment]
On track

**Comment by chris-kuryak** [comment]
@lucero-v Add results/summary of your findings to the ticket.
This will then be reviewed at Engr sync, to determine a potential solution and a follow-on ticket.

**Comment by lucero-v** [comment]
## Findings
appinsights -> Performance   
Showing data from the past year
<img width="1169" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMjB2Q0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--9b963d860bdda14e48df4ce0cf68be943a420d5d/image.png" alt="image.png" />
<img width="651" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBM0V2Q0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--482f440bf9b7e74701e60b54b8f2f9d7ec5a9582/image.png" alt="image.png" />

From the first graph we can tell that the send step takes on average 7.42s which is less than the 9.09s it takes for the convert-fhir step on the pipeline. Furthermore, the send step duration includes an external dependency duration when we submit via REST. On the second graph we can see that sending reports to OK takes significantly longer compared to CA.

The following graph is from the Metrics screen, showing data from 11/1/24 to 12/17/24 :

<img width="914" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMzR2Q0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--ef110ea61534a8799c3365f77c13ecc46899eb44/image.png" alt="image.png" />

One thing to note here are the spikes in send duration. I looked into them and found that those happened whenever the send step ran into an exception or warning. Also note that the sendSuccesses count is the same as the sendFailures count, which does not make sense but I don't have the full context on where "sendFailure" and "sendSuccess" are being defined. This led to me dig into a different source: logs and customEvents. Through the logs I was able to run some queries to get a breakdown of a successful send step duration per receiver. Below are the results: 

<img width="1429" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNEF2Q0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--818272fae05c40e22e5a015b044f8c5137ce531d/image.png" alt="image.png" />

Note that the OK receivers take the longest which makes sense because the http calls to their endpoint take about a minute to complete.

I also ran a query to get the durations whenever the send step encounters an issue and found that it correlated to the spikes that we saw in the Metrics line graph.

<img width="960" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNEl2Q0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--e3f19cbc450f49d03e11dc6d8fdb21d4e813ef8d/image.png" alt="image.png" />

#### Relevant Queries:
```
let reportSentEvents = (customEvents 
    | where name == "REPORT_SENT"
    | project operation_ParentId, eventCustomDimensions = customDimensions
);
requests
| where timestamp > datetime("2023-12-16T00:01:00.000Z") and timestamp < datetime("2024-12-16T00:01:00.000Z")
| where name in ("send")
| join reportSentEvents
    on $left.id == $right.operation_ParentId
| project timestamp, duration, eventCustomDimensions
| extend durationMins = round(duration/60000,2), params = parse_json(tostring(eventCustomDimensions.params)), reports = parse_json(tostring(eventCustomDimensions.submittedReportIds))
| extend receiverName = tostring(params.receiverName)
    ,transportType = tostring(params.transportType)
    ,reportsCount = array_length(reports)
| summarize requestCount = count(), avgDurationMins = round(avg(durationMins),2), maxDurationMins = max(durationMins), avgReportCount = round(avg(reportsCount),2), maxReportCount = max(reportsCount), transportType = max(transportType) by receiverName
| order by avgDurationMins desc
```

```
let problematicTraces = (traces 
    | where severityLevel > 1
    | project operation_ParentId, traceCustomDimensions = customDimensions, message
);
requests
| where timestamp > datetime("2023-12-16T00:01:00.000Z") and timestamp < datetime("2024-12-16T00:01:00.000Z")
| where name in ("send") and duration > 60000
| join problematicTraces
    on $left.id == $right.operation_ParentId
| project timestamp, durationMs = duration, traceCustomDimensions, message
| extend durationMins = round(durationMs/60000,2)
| order by durationMs desc```

**Comment by lucero-v** [comment]
## Takeaway

The concern with the send step might have been misled since we need to consider the external dependency of the receivers. If anything, we need to improve the way we handle exceptions and maybe add a custom event to track retries/unsuccessful sends.

**Comment by chris-kuryak** [comment]
Waiting on Platform to review. @arnejduranovic to set up time when he returns from OOO. (Let me know if this is incorrect @arnejduranovic )

**Comment by chris-kuryak** [comment]
Lucero planning on reviewing with Platform team at MIchael's office hours on Fri 1/10/25. We will know how to proceed after that conversation.

**Comment by lucero-v** [comment]
Discussion takeaways from Michael's office hours:
https://cdc-my.sharepoint.com/:w:/r/personal/xwc4_cdc_gov/Documents/Send%20Step%20Analysis.docx?d=wb66ad5e4412d4754a82dbb7b589b2bc4&csf=1&web=1&e=ka8A4f
